storybookjs/storybook
# Svelte: Fix decorators always running twice

Closes #24549 






## What I did

Inside PreviewRender.svelte storyFn was called twice, once to get the values and once in the reactive label statement to make it updated whenever storyFn changes. I've added a small utility

```ts
// the first time we don't want to call storyFn two times so we just return the values
// we already have from the previous call. If storyFn changes this function will run
// again but this time firstTime will be false
function getStoryFnValue(storyFn){
	if(firstTime){
		firstTime = false;
		return {
			Component,
			props,
			on,
		}
	}
	return storyFn();
}
```
that the first time return the values already present by the first call to storyFn.

## Checklist for Contributors

### Testing



#### The changes in this PR are covered in the following automated tests:

#### Manual testing

I tested that decorators only run once. You can verify that by opening one of the svelte sandbox and by going to `renderers/svelte/decorators-runs-once` verifying that in the console there's only one `decorator` log.



### Documentation



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release



This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_
